orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -207,39 +280,274 @@ Status TrainingRunner::Run(IDataLoader* training_data_loader, IDataLoader* test_
+    ORT_ENFORCE(params_.num_pipeline_stages > 1);
+    feed_names.push_back(pipeline_context_.forward_recorded_event_name);
+    OrtValue event_id;
+    const int64_t id = pipeline_schedule_.get_forward_recorded_event_id(
+      static_cast<int>(pipeline_context_.pipeline_stage_id),
+      static_cast<int>(step_ % pipeline_context_.num_gradient_accumulation_steps));
+    TrainingUtil::CreateCpuMLScalar(
+      id,
+      &event_id,
+      input_allocator_);
+    feeds.push_back(event_id);
+  }
+
+  // Create feed of waited event in backward pass.
+  if (!pipeline_context_.backward_waited_event_name.empty()) {
+    ORT_ENFORCE(params_.num_pipeline_stages > 1);
+    feed_names.push_back(pipeline_context_.backward_waited_event_name);
+    OrtValue event_id;
+    const int64_t id = pipeline_schedule_.get_backward_waited_event_id(
+      static_cast<int>(pipeline_context_.pipeline_stage_id),
+      static_cast<int>(step_ % pipeline_context_.num_gradient_accumulation_steps));
+    TrainingUtil::CreateCpuMLScalar(
+      id,
+      &event_id,
+      input_allocator_);
+    feeds.push_back(event_id);
+  }
+
+  // Create feed of recorded event in backward pass.
+  if (!pipeline_context_.backward_recorded_event_name.empty()) {
+    ORT_ENFORCE(params_.num_pipeline_stages > 1);
+    feed_names.push_back(pipeline_context_.backward_recorded_event_name);
+    OrtValue event_id;
+    int64_t id = pipeline_schedule_.get_backward_recorded_event_id(
+      static_cast<int>(pipeline_context_.pipeline_stage_id),
+      static_cast<int>(step_ % pipeline_context_.num_gradient_accumulation_steps));
+    TrainingUtil::CreateCpuMLScalar(
+      id,
+      &event_id,
+      input_allocator_);
+    feeds.push_back(event_id);
+  }
+
+  return Status::OK();
+}
+
+Status TrainingRunner::PrepareFetchNamesAndFetches(const bool do_weight_update,
+                                                   std::vector<std::string>& fetch_names,
+                                                   std::vector<MLValue>& fetches) {
+  // Initialize outputs of this function.
+  fetch_names = std::vector<std::string>();
+  fetches = std::vector<MLValue>();
+
+  const auto& allowed_fetch_names = pipeline_context_.fetch_names;
+
+  if (do_weight_update) {
+    // Set up tensor to be fetched when doing model update. 
+
+    if (params_.num_pipeline_stages > 1) {
+      // If pipeline is used, we need to filter out fetches which are not in this pipeline stage.
+
+      for (size_t i = 0; i < params_.fetch_names.size(); ++i) {
+        const auto name = params_.fetch_names[i];
+        auto it = std::find(allowed_fetch_names.begin(), allowed_fetch_names.end(), name);
+        if (it == allowed_fetch_names.end()) {
+          continue;
+        }
+        fetch_names.push_back(name);
+      }
+    } else {
+      // No pipeline. All fetched names should appear in the graph handled by this process.
+      fetch_names = params_.fetch_names;
```

Comment:
Seems no need to special case if allowed_fetch_names contain only fetch for given pipeline stage.

Reply:
`params_fetch_names` is actually user-specified tensor names. Because not all fetchable tensors would be fetched, we have this
```
else {
      // No pipeline. All fetched names should appear in the graph handled by this process.
      fetch_names = params_.fetch_names;
```
to only fetch tensors user wants.

Reply:
What I mean is that if we always set up allowed_fetch_names, regardless of using pipeline or not, then we just need the logic inside if (params_.pipeline_parallel_size > 1), and no need to check pipeline parallel size. Could there be other cases that allowed_fetch_names not matching params_.fetch_names, when not using pipeline?

---
In reply to: [419608711](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r419608711) [](ancestors = 419608711)